### DISCUSSION TITLE
choose with keyboard shortcut

### DISCUSSION THREAD

**Discussion by SevereOverfl0w** [charmbracelet/gum#847] [Ideas]
I'd love a way to make "quick menus" with gum where each item corresponds to a keyboard shortcut of some kind for it's selection.  This way I could type "a" or "b" or "c" to quickly choose between 3 presented options.  On inputting this character (no need to hit enter) it would be selected as the choice.

**Comment by mathomp4** [comment]
I want to add my +1 here. I'm using `gum choose` in a script and can often get like 20 or more items in the choices. If I could see:

```
❯ gum choose "alpha" "beta" "gamma" "delta" ... "chi" "psi" "omega"
> 1. alpha
  2. beta
  3. gamma
  4. delta
...
  22. chi
  23. psi
  24. omega
```
or some such and type "22" to get to chi rather than scrolling allll the way down, that'd be great!

Of course, the "no need for enter" in @SevereOverfl0w's case wouldn't work for my case as if I want #22, it'd go to #2 and be done. Maybe a "more than 9 items means you have to type and press enter"?

(Though it still won't stop me from trying to select it with my mouse about once a week!)

**Comment by maaslalani** [comment]
This is a great idea @SevereOverfl0w!

I agree with @mathomp4, where if the options are > 10 then we should keep a small buffer of the number keys pressed and navigate to that option instead of selecting it.

> Though it still won't stop me from trying to select it with my mouse about once a week

I believe mouse support is possible with Bubble Tea, this can also be something we add.

**Comment by keevan** [comment]
+1, this would be very useful.

It'd be the alternative for a gum filter but give you more than the current `choose`. I'm thinking it would be nice to have it display the shortcuts at the bottom of the selection (similar to what gum pager does currently)
![image](https://user-images.githubusercontent.com/9924643/195290635-28f8f391-2d7a-4b2e-b90b-6833b70f2bf9.png)

I'm not sure if this approach would work as well for long lists that aren't triggered with a single character, such as @mathomp4's example. I would personally pick up `gum filter` in that example instead, along with an enhancement similar to this https://github.com/charmbracelet/gum/issues/128 but for filter elements.

**Comment by mathomp4** [comment]
@keevan I think my `gum filter` version of my code might be fixed by #186 if it works. I am usually typing whole words in `filter` code anyway. 

And, yes, I still try to use my mouse in both `choose` and `filter` like once a week. 😆

**Comment by edrex** [comment]
I'm interested in the use case of explicitly binding keys to a small number of items. I imagine it could look something like this:

```
cat | gum choose --keys
x$Exercise
c$Conglomerate
p$Profit
^d
> Exercise (x)
  Conglomerate (c)
  Profit (p)
```

(not sure what the most appropriate delimiter would be for inputs)

**Comment by crvillanueva** [comment]
I currently do this with bash scripts, it would be amazing if I could do the same with gum. I hope this feature is added in the future.

**Comment by edrex** [comment]
I currently use [tydra](https://github.com/Mange/tydra) for this, but it's unmaintained.

**Comment by caarlos0** [comment]
this would probably be a `gum menu` kind of thing, I think

**Comment by abitrolly** [comment]
`gum menu` would be nice, because in `menu` you want to separate label to choose from, and commands that will be executed as a result.

**Comment by meowgorithm** [comment]
Hey everyone! This is a nice feature, however I'm going to convert it into a discussion (for further discussion) until we've decided to integrate it. In general, we're aiming to be as thoughtful as possible about increasing Gum's surface area given our internal capacity.

As an aside, building a custom menu like this would be relatively trivial in Bubble Tea (the [intro tutorial](https://github.com/charmbracelet/bubbletea/tree/main/tutorials/basics) is actually a great starting place).

**Comment by abitrolly** [comment]
I made a menu for selecting `git branch` as a OMZ plugin. There are some bits that can be generalized, but the tricky parts are:

1. process input line to get (value, menuline) and then run command with value
2. highlighting menu items, because need to parse ANSI state

![image](https://github.com/user-attachments/assets/bcff2be3-4a3d-4bc4-8b30-1316ae2b9fd2)

https://github.com/ohmyzsh/ohmyzsh/discussions/12993
